### PR TITLE
41926: Refactory proposal

### PR DESCRIPTION
**Pull request by xiangzhai**
https://github.com/dotnet/runtime/issues/41926

It is just the skeleton to improve readability and makes it easier to port to different platforms.

Welcome feedback!

Thanks,
Leslie Zhai

### CODE REVIEW

**Comment by jkotas** [comment]
cc @BruceForstall

File: src/coreclr/src/jit/morpharmarch.cpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+#include "jitpch.h"
+#ifdef _MSC_VER
+#pragma hdrstop
+#endif
+
+#include "allocacheck.h" // for alloca
+
+void Compiler::fgInitArgInfoHelp(GenTreeCall* call)
+{
```
Comment (jkotas) [inline src/coreclr/src/jit/morpharmarch.cpp:12]: I do not see what code would move here from `fgInitArgInfo` to improve readability and maintainability of the code.

Could you please include the actual code that you expect to be moved here?

Comment (xiangzhai) [reply src/coreclr/src/jit/morpharmarch.cpp:12]: Thanks for your review!

I included the actual code.

Please review the skeleton again.

Thanks,
Leslie Zhai

**Comment by xiangzhai** [comment]
> runtime (CoreCLR Pri0 Runtime Tests Run Linux arm checked) Failing after 31m — CoreCLR Pri0 Runtime Tests Run Linux arm checked failed

False positive?

**Comment by jkotas** [comment]
It is a real crash that the change introduced. It demonstrates that refactorings like this come with a real risk of introducing regressions.

**Comment by jkotas** [comment]
The failure is:
```
Assert failure(PID 3136 [0x00000c40], Thread: 3136 [0x0c40]): Assertion failed 'argCount < argTableSize' in 'ILStubClass:IL_STUB_WrapperDelegate_Invoke(System.__Canon,System.__Canon):int:this' during 'Morph - Global' (IL size 14)
```

**Comment by xiangzhai** [comment]
Hi @jkotas 

Sorry for the regression.

I can't not `./build.sh -subset clr` now:  https://github.com/dotnet/runtime/issues/33294

So I need the CI for the regression test...

How to workaround the build?  And I need to build the testcase in my local PC.  Then I could do the test in my local enviroment at first.

Thanks,
Leslie Zhai

**Comment by xiangzhai** [comment]
> The failure is:
> 
> ```
> Assert failure(PID 3136 [0x00000c40], Thread: 3136 [0x0c40]): Assertion failed 'argCount < argTableSize' in 'ILStubClass:IL_STUB_WrapperDelegate_Invoke(System.__Canon,System.__Canon):int:this' during 'Morph - Global' (IL size 14)
> ```

Thank you so much!

File: src/coreclr/src/jit/morpharmarch.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            GenTreeAddrMode(TYP_BYREF, arg, nullptr, 0, eeGetEEInfo()->offsetOfWrapperDelegateIndirectCell);
+
+        // Append newArg as the last arg
+        GenTreeCall::Use** insertionPoint = &call->gtCallArgs;
+        for (; *insertionPoint != nullptr; insertionPoint = &((*insertionPoint)->NextRef()))
+        {
+        }
+        *insertionPoint = gtNewCallArgs(newArg);
+
+        (*numArgs)++;
```
Comment (xiangzhai) [inline src/coreclr/src/jit/morpharmarch.cpp:56]: Sorry for my stupid fault!

**Comment by xiangzhai** [comment]
> runtime (Formatting Linux x64) Failing after 5m — Formatting Linux x64 failed

How to fix the formatting issue?

**Comment by BruceForstall** [comment]
To fix formatting, you can either follow the instructions in the formatting job output to download and apply a patch:

https://dev.azure.com/dnceng/public/_build/results?buildId=812582&view=logs&jobId=7b44e949-88ac-5e54-d61f-6dbb52b4ac02&j=7b44e949-88ac-5e54-d61f-6dbb52b4ac02&t=45872f7a-05a6-54b6-3508-a9933282fd65

Or, in general, you can run `jit-format` as described here: https://github.com/dotnet/jitutils. (This sometimes doesn't fix formatting issues that would only show up on other platforms.)

**Comment by xiangzhai** [comment]
Hi @BruceForstall 

Thanks for your teaching!

> To fix formatting, you can either follow the instructions in the formatting job output to download and apply a patch:
> 
> https://dev.azure.com/dnceng/public/_build/results?buildId=812582&view=logs&jobId=7b44e949-88ac-5e54-d61f-6dbb52b4ac02&j=7b44e949-88ac-5e54-d61f-6dbb52b4ac02&t=45872f7a-05a6-54b6-3508-a9933282fd65

Perhaps my azure account can not view the 'Build artifacts published' entry. 

> 
> Or, in general, you can run `jit-format` as described here: https://github.com/dotnet/jitutils. (This sometimes doesn't fix formatting issues that would only show up on other platforms.)

I fixed the formatting issue by `jit-format -a x64 -b Checked -o Linux -c /run/media/zhaixiang/data/project/xiangzhai/runtime-41926/src/coreclr --verbose --fix --projects dll`.

Please review my patch again.

Thanks,
Leslie Zhai

File: src/coreclr/src/jit/morph.cpp [hunk 3]
```diff
@@ -2668,6 +2504,9 @@ void Compiler::fgInitArgInfo(GenTreeCall* call)
         noway_assert(arg != nullptr);
         call->gtCallCookie = nullptr;
 
+        // TODO: Refactory proposal
+        // I am not good at naming.  The function name might be fgInitArgInfoHelpN below.
+
 #if defined(TARGET_X86)
```
Comment (xiangzhai) [inline src/coreclr/src/jit/morph.cpp:2510]: The patch is only the skeleton.

It needs to choose the precise name for the functions, but not the `fgInitArgInfoHelp` or `fgInitArgInfoHelpN` to replace the `#ifdef _TARGET_YOUR_ARCH` in the morph.cpp.

Please give me some advice about the naming.

Thanks,
Leslie Zhai

**Comment by xiangzhai** [comment]
> runtime (Formatting Linux x64) Failing after 5m — Formatting Linux x64 failed

`clang-format` old version issue?

Because ArchLinux, my local enviroment, clang-format version is 10.0.1.

**Comment by xiangzhai** [comment]
@Maoni0   But I used `jit-format` and `git diff` to check my patch.  There is no formatting issue...

Please give me some hint.

Thanks,
Leslie Zhai

**Comment by ViktorHofer** [comment]
// Auto-generated message

https://github.com/dotnet/runtime/commit/69e114c1abf91241a0eeecf1ecceab4711b8aa62 which was merged 12/7 removed the intermediate src/coreclr/**src**/ folder. This PR needs to be updated as it touches files in that directory which causes conflicts.

To update your commits you can use this bash script: https://gist.github.com/ViktorHofer/6d24f62abdcddb518b4966ead5ef3783. Feel free to use the comment section of the gist to improve the script for others.

**Comment by xiangzhai** [comment]
Hi @ViktorHofer 

> 69e114c which was merged 12/7 removed the intermediate src/coreclr/src/ folder

Great!  It just like OpenJDK 8u migrated to master, changed the hotsport/src/cpu/yourarch/vm to src/hotspot/cpu/yourarch folder.

Thanks,
Leslie Zhai

**Comment by BruceForstall** [comment]
@xiangzhai Thanks for your efforts to improve the readability and ease or programming in the JIT code base. My apologies for the slow response. As you've discovered, fgMorphArgs is very complicated and full of #ifdef. What is the scope of your effort here? You've suggested a start here. For this particular PR, what do you want to accomplish? Do you want to extract all the non-standard argument handling? (Maybe name it fgInitNonStandardArgInfo?) This would eliminate code and #ifdefs from the main function, but would possibly introduce some duplication of code elsewhere, so it's not clear that's a good trade-off, if there ends up being duplicated code. A simple start might be to just hoist all the non-standard arg setup to its own function, leaving all the #ifdef intact. Although that wouldn't accomplish too much.

(btw, you need to remove the extra "src" directory in your morpharmarch.cpp/morphxarch.cpp files)

**Comment by xiangzhai** [comment]
Hi @BruceForstall 

> My apologies for the slow response.

You are welcome.  And I am not available after February 8th due to Chinese spring festival.

> but would possibly introduce some duplication of code elsewhere

Yes, there is the same issue in OpenJDK, for example, [SharedRuntime::generate_i2c2i_adapters](https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp#L946) to generate `Interpreter` to `Compiler` or `Compiler` to `Interpreter` adapters.

For x86:

```
AdapterHandlerEntry* SharedRuntime::generate_i2c2i_adapters(MacroAssembler *masm,
                                                            int total_args_passed,
                                                            int comp_args_on_stack,
                                                            const BasicType *sig_bt,
                                                            const VMRegPair *regs,
                                                            AdapterFingerPrint* fingerprint) {
  address i2c_entry = __ pc();                                                  
                                                                                
  gen_i2c_adapter(masm, total_args_passed, comp_args_on_stack, sig_bt, regs);   
                                                                                
  // -------------------------------------------------------------------------  
  // Generate a C2I adapter.  On entry we know rbx holds the Method* during calls
  // to the interpreter.  The args start out packed in the compiled layout.  They
  // need to be unpacked into the interpreter layout.  This will almost always  
  // require some stack space.  We grow the current (compiled) stack, then repack
  // the args.  We  finally end in a jump to the generic interpreter entry point.
  // On exit from the interpreter, the interpreter will restore our SP (lest the
  // compiled code, which relys solely on SP and not RBP, get sick).            
                                                                                
  address c2i_unverified_entry = __ pc();                                       
  Label skip_fixup;                                                             
  Label ok;                                                                     
                                                                                
  Register holder = rax;                                                        
  Register receiver = j_rarg0;                                                  
  Register temp = rbx;                                                          
                                                                                
  {                                                                             
    __ load_klass(temp, receiver, rscratch1);                                   
    __ cmpptr(temp, Address(holder, CompiledICHolder::holder_klass_offset()));  
    __ movptr(rbx, Address(holder, CompiledICHolder::holder_metadata_offset()));
    __ jcc(Assembler::equal, ok);                                               
    __ jump(RuntimeAddress(SharedRuntime::get_ic_miss_stub()));                 
                                                                                
    __ bind(ok);
...
```

For aarch64:

```
AdapterHandlerEntry* SharedRuntime::generate_i2c2i_adapters(MacroAssembler *masm,
                                                            int total_args_passed,
                                                            int comp_args_on_stack,
                                                            const BasicType *sig_bt,
                                                            const VMRegPair *regs,
                                                            AdapterFingerPrint* fingerprint) {
  address i2c_entry = __ pc();                                                  
                                                                                
  gen_i2c_adapter(masm, total_args_passed, comp_args_on_stack, sig_bt, regs);   
                                                                                
  address c2i_unverified_entry = __ pc();                                       
  Label skip_fixup;                                                             
                                                                                
  Label ok;                                                                     
                                                                                
  Register holder = rscratch2;                                                  
  Register receiver = j_rarg0;                                                  
  Register tmp = r10;  // A call-clobbered register not used for arg passing    
                                                                                
  // -------------------------------------------------------------------------  
  // Generate a C2I adapter.  On entry we know rmethod holds the Method* during calls
  // to the interpreter.  The args start out packed in the compiled layout.  They
  // need to be unpacked into the interpreter layout.  This will almost always  
  // require some stack space.  We grow the current (compiled) stack, then repack
  // the args.  We  finally end in a jump to the generic interpreter entry point.
  // On exit from the interpreter, the interpreter will restore our SP (lest the
  // compiled code, which relys solely on SP and not FP, get sick).             
                                                                                
  {                                                                             
    __ block_comment("c2i_unverified_entry {");                                 
    __ load_klass(rscratch1, receiver);                                         
    __ ldr(tmp, Address(holder, CompiledICHolder::holder_klass_offset()));          
    __ cmp(rscratch1, tmp);                                                     
    __ ldr(rmethod, Address(holder, CompiledICHolder::holder_metadata_offset()));
    __ br(Assembler::EQ, ok);                                                   
    __ far_jump(RuntimeAddress(SharedRuntime::get_ic_miss_stub()));             
                                                                                
    __ bind(ok);
...            
```

> A simple start might be to just hoist all the non-standard arg setup to its own function, leaving all the #ifdef intact

How about just add `fgInitArgInfoHelp` in the src/coreclr/jit/morph.cpp ?   Sorry that I am not good at naming.

```
void Compiler::fgInitArgInfoHelp(GenTreeCall* call, NonStandardArgs* nonStandardArgs, unsigned* numArgs)
{
#if defined(TARGET_X86) || defined(TARGET_ARM)
...
#endif // defined(TARGET_X86) || defined(TARGET_ARM)
#if defined(TARGET_ARM)
...
#endif
#if defined(TARGET_YOUR_ARCH)
...
#endif
}
```

Happy [NIU](https://en.wikipedia.org/wiki/Ox_(zodiac)) year,
Leslie Zhai

**Comment by BruceForstall** [comment]
I am going to close this PR. The ideas here are worth pursuing, but require time and attention and probably a broader scope than this. @xiangzhai it doesn't seem like you are focused on this now, and it's not high on our priority list, either. We will certainly consider the ideas going forward (I was working on code in this area just yesterday).

**Pull request closed without merging**